### PR TITLE
Add username + password authentication

### PR DESCRIPTION
**Pull request by duncte123**
Tasks
- [x] Add password to profile
- [x] Verifiy emails
- [x] Add login routes
- [x] Add register routes
- [x] Add 2fa
- [x] Add disable 2fa
- [x] Add password resets

Notes: 
- https://www.twilio.com/blog/html-attributes-two-factor-authentication-autocomplete

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -27,11 +27,16 @@ services:
       # sparkly new oauth2
       - TWITTER_OAUTH2_CLIENT_ID=
       - TWITTER_OAUTH2_CLIENT_SECRET=
+      - HIKARI_MAXIMUM_POOL_SIZE=1000
```
Comment (MaxieTheHatter) [inline docker-compose.yml:30]: 1000 is a lot of concurrent conenctions, do you really need that many?
https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing has some interesting reading material

File: src/main/java/app/oengus/entity/model/EmailVerification.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package app.oengus.entity.model;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import javax.persistence.*;
```
Comment (MaxieTheHatter) [inline src/main/java/app/oengus/entity/model/EmailVerification.java:6]: Star imports are considered code smell, tip is to set your IDE to only do star imports at a very high number (99 or higher).

File: src/main/java/app/oengus/entity/model/User.java [hunk 3]
```diff
@@ -136,7 +159,7 @@ public boolean isAccountNonExpired() {
     @Override
     @JsonIgnore
     public boolean isAccountNonLocked() {
-        return false;
+        return !roles.contains(Role.ROLE_BANNED);
     }
```
Comment (MaxieTheHatter) [inline src/main/java/app/oengus/entity/model/User.java:163]: Double negation, following would reduce cognitive complexity:
```
public boolean isAccountLocked() {
    return roles.contains(Role.ROLE_BANNED);
}
```

Comment (duncte123) [reply src/main/java/app/oengus/entity/model/User.java:163]: Unfortunately, this method comes from the spring framework. I can add that second method but this override is required

**Review by MaxieTheHatter — COMMENTED** [review]
Some nitpicks, some good practices.

How do you handle the case when a user has a valid jwt but accesses a resource they shouldn't?

**Comment by duncte123** [comment]
> Some nitpicks, some good practices.
> 
> How do you handle the case when a user has a valid jwt but accesses a resource they shouldn't?

A good example of that is visible in the v1 api. https://github.com/esamarathon/oengusio/blob/master/src/main/java/app/oengus/web/v1/MarathonController.java#L169

The `@PreAuthorize` annotation uses methods defined in the [CustomMethodSecurityExpressionRoot](https://github.com/esamarathon/oengusio/blob/master/src/main/java/app/oengus/spring/security/CustomMethodSecurityExpressionRoot.java#L92-L106) class. If the method returns true the code from the controller will be called, otherwise the user is presented with a 403 response.

**Comment by MaxieTheHatter** [comment]
> > Some nitpicks, some good practices.
> > How do you handle the case when a user has a valid jwt but accesses a resource they shouldn't?
> 
> A good example of that is visible in the v1 api. https://github.com/esamarathon/oengusio/blob/master/src/main/java/app/oengus/web/v1/MarathonController.java#L169
> 
> The `@PreAuthorize` annotation uses methods defined in the [CustomMethodSecurityExpressionRoot](https://github.com/esamarathon/oengusio/blob/master/src/main/java/app/oengus/spring/security/CustomMethodSecurityExpressionRoot.java#L92-L106) class. If the method returns true the code from the controller will be called, otherwise the user is presented with a 403 response.

All good then, it's important to not return a 401 for stuff like that as it invalidates the JWT

**Review by MaxieTheHatter — APPROVED** [review]

**Pull request merged**